accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,126 @@
+
+A developer has confidence that their library using source generators will continue to work in the same way a library without source generators works on future versions of .NET following the same compatibility rules.
+
+## Requirements
+
+### Goals
+
+- Enable consumption of source generators for components in shared frameworks
+- Address conflict resolution between source generators in shared frameworks and nuget packages
+- Define servicing process for shared framework source generators
+- Establish guidelines building and versioning shared framework source generators
+- Describe compatibility requirements for source generators.
+
+### Non-Goals
+
+- Define a process for components not in shared frameworks to contribute source generators.
+- Define any changes to the source generator architecture.
+- Define any implementation details of source generators or source generator tests.
+- Define a process around internal or non-shipping source generators
+
+## Stakeholders and Reviewers
+
+- ASP.NET - also building shared framework source generators.
+- .NET Libraries team - building multiple shared framework source generators.
+- Mono team - interested in providing source generators.
+- Roslyn team - responsible for source generator infrastructure.
+- .NET SDK team - responsible for resolving shared frameworks and doing conflict resolution with content.
+- NuGet team - FYI only, no changes requested
+
+## Design
+
+### Framework provided source generators
+
+Framework provided source generators should have a mechanism to be exposed and versioned along side the framework and library which they extend.  These source generators will be passed to the compiler when that framework is targeted.  These source generators will not ne used when the framework is not targeted.  For example, ASP.NET specific source generators will not be used when ASP.NET is not referenced.  The version of the analyzers specific to the TargetFramework version will be used.  For example, when targeting .NET 6.0 the analyzers specific to net6.0 will be used and not those for .NET 7.0, even when using the SDK that ships with .NET 7.0.
+
+To facilitate this, we will package analyzers inside [reference packs](https://github.com/dotnet/designs/blob/main/accepted/2019/targeting-packs-and-runtime-packs.md).  We will use the standard [nuget conventions for describing analyzers](https://github.com/dotnet/designs/blob/main/accepted/2019/targeting-packs-and-runtime-packs.md).  The .NET SDK will be responsible for locating the appropriate analyzers for the TargetFramework and creating `@(Analyzer)` items as part of the `ResolveTargetingPackAssets` task.  These items will be conflict-resolved with other Analyzer sources based as described in the following section.
```

Comment:
We don't use the NuGet conventions to interpret the contents of targeting packs.  The contents of a targeting pack are described by a `data/FrameworkList.xml` file.  The paths that the files are actually found under may (or may not) conform to the NuGet conventions.

So we will need to define how analyzers are included in targeting packs.  I would go with either adding them to the FrameworkList.xml file, or possibly creating a new file specific for analyzers.

Reply:
We do already follow nuget convention for refs.  I would prefer we follow the same convention for analyzers, even if the actual resolution is more of a simple path check.  That way we don't "forget" about some constraining variable like language.

I thought we used a wild-card for refs but now I see that the SDK constructs all the paths from information in the framework list.
https://github.com/dotnet/sdk/blob/8659ce5f7550359048083a1ddafb25f105d269b8/src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs#L58

Is this a perf optimization to avoid hitting the disk?

Reply:
Ok, I think that's the case.  I'll address this in the doc.

Reply:
Yes, it's partly about perf.  I'm not sure enumerating the files on disk would impact perf, but we also store the assembly and file versions in the framework list which helps.

It's also helpful to rely on the framework list for the file path instead of conventions, so we don't have to worry about whether we should use net5.0 or net5.0-windows (or netcoreapp5.0) as part of the path, for example.  We already used the target framework to select the targeting pack, so using it as part of a convention for the path inside the pack just complicates things.

For analyzers, rather than re-implement the NuGet convention (which actually [isn't currently implemented in NuGet](https://github.com/NuGet/Home/issues/6279)), I think it's better to just have them listed in a file such as the framework list.